classes/object_file_system.php

```diff
@@ -492,15 +513,27 @@ public function delete_object_from_hash($contenthash) {
 
         switch ($location) {
             case OBJECT_LOCATION_LOCAL:
+                $this->copy_local_file_to_trashdir_from_hash($contenthash);
                 $this->delete_local_file_from_hash($contenthash);
                 break;
+
             case OBJECT_LOCATION_DUPLICATED:
                 $this->delete_local_file_from_hash($contenthash);
-                $this->delete_external_file_from_hash($contenthash);
+                $success = $this->copy_external_file_to_trashdir_from_hash($contenthash);
+                if ($success) {
+                    $this->delete_external_file_from_hash($contenthash);
+                }
                 break;
+
             case OBJECT_LOCATION_EXTERNAL;
-                $this->delete_external_file_from_hash($contenthash);
+                $success = $this->copy_external_file_to_trashdir_from_hash($contenthash);
+                if ($success) {
+                    $this->delete_external_file_from_hash($contenthash);
+                }
                 break;
+
+            case OBJECT_LOCATION_ERROR;
+                return;
```

Comment:
I'd recommend to add default option  as well here.

Reply:
Added in further commit.